src/PythonMigrationViewExtension/PythonMigrationViewExtension.cs

```diff
@@ -163,49 +174,219 @@ private void OnNodeAdded(Graph.Nodes.NodeModel obj)
         {
-            if (!(obj is PythonNodeBase pythonNode))
-                return;
-
-            UnSubscribePythonNodeEvents(pythonNode);
+            if (obj is PythonNodeBase pythonNode)
+            {
+                UnSubscribePythonNodeEvents(pythonNode);
+            }
+            else if (obj is Dynamo.Graph.Nodes.CustomNodes.Function customNode)
+            {
+                var cnm = DynamoViewModel?.Model?.CustomNodeManager;
+                var defId = customNode.Definition?.FunctionId ?? Guid.Empty;
+                if (defId == Guid.Empty || cnm == null) return;
+
+                // Remove tracking if this was the last instance of that custom node
+                var anyRemaining = CurrentWorkspace?.Nodes
+                    .OfType<Dynamo.Graph.Nodes.CustomNodes.Function>()
+                    .Any(f => (f.Definition?.FunctionId ?? Guid.Empty) == defId) == true;
+
+                if (!anyRemaining
+                    && cnm.TryGetFunctionWorkspace(defId, Models.DynamoModel.IsTestMode, out ICustomNodeWorkspaceModel defWsModel) == true
+                    && defWsModel is WorkspaceModel defWs)
+                {
+                    upgradeService.TouchedCustomWorkspaces.Remove(defWs);
+                    upgradeService.CustomToastShownDef.Remove(defId);
+                }
+            }
         }
 
         private void OnCurrentWorkspaceChanged(IWorkspaceModel workspace)
         {
             UnSubscribeWorkspaceEvents();
+
+            var previous = CurrentWorkspace;
             CurrentWorkspace = workspace as WorkspaceModel;
             PythonDependencies.UpdateWorkspace(CurrentWorkspace);
+
             SubscribeToWorkspaceEvents();
 
-            NotificationTracker.Remove(CurrentWorkspace.Guid);
+            // Fresh per-workspace notification state
+            CurrentWorkspace.HasShownCPythonNotification = false;
+
+            // Close any upgrade toast when switching tabs/workspaces.
+            DynamoViewModel.MainGuideManager?.CloseRealTimeInfoWindow();
+
+            if (previous != null)
+            {
+                NotificationTracker.Remove(previous.Guid);
+            }
             GraphPythonDependencies.CustomNodePythonDependencyMap.Clear();
 
-            RecomputeCPython3NotificationForWorkspace();
+            if (CurrentWorkspace is HomeWorkspaceModel hws)
+            {
+                if (lastWorkspaceGuid != hws.Guid)
+                {
+                    // Track last home workspace to avoid duplicate work on the same one
+                    lastWorkspaceGuid = hws.Guid;
+                    bool isAutoRun = hws.RunSettings.RunType == RunType.Automatic;
+
+                    // If opening in Automatic, switch to Manual before upgrading to avoid mutating during eval
+                    if (isAutoRun)
+                    {
+                        hws.RunSettings.RunType = RunType.Manual;
+                    }
```

Comment:
Although unlikely, `hws.RunSettings.RunType` can also be `RunType.Periodic`, would you think this might be slightly better?

```cs
var oldRunType = hws.RunSettings.RunType;
hws.RunSettings.RunType = RunType.Manual; // Always set it to `manual`

RecomputeCPython3NotificationForWorkspace();

// Always restore the run type
System.Windows.Application.Current.Dispatcher.BeginInvoke(
    () => hws.RunSettings.RunType = oldRunType,
    System.Windows.Threading.DispatcherPriority.ApplicationIdle);
```